watertap/property_models/IX_prop_pack.py

```diff
@@ -0,0 +1,1110 @@
+    Param,
+    Set,
+    Suffix,
+    value,
+    check_optimal_termination,
+    units as pyunits,
+)
+
+from pyomo.common.config import ConfigValue, In
+
+# Import IDAES cores
+from idaes.core import (
+    declare_process_block_class,
+    MaterialFlowBasis,
+    PhysicalParameterBlock,
+    StateBlockData,
+    StateBlock,
+    MaterialBalanceType,
+    EnergyBalanceType,
+)
+from idaes.core.components import Component, Solute, Solvent, Cation, Anion
+from idaes.core.phases import LiquidPhase, AqueousPhase
+from idaes.core.util.constants import Constants
+from idaes.core.util.initialization import (
+    fix_state_vars,
+    revert_state_vars,
+    solve_indexed_blocks,
+)
+from idaes.core.util.misc import add_object_reference, extract_data
+from idaes.core.util import get_solver
+from idaes.core.util.model_statistics import (
+    degrees_of_freedom,
+    number_unfixed_variables,
+)
+from idaes.core.util.exceptions import (
+    ConfigurationError,
+    InitializationError,
+    PropertyPackageError,
+)
+import idaes.core.util.scaling as iscale
+
+# Set up logger
+_log = idaeslog.getLogger(__name__)
+
+
+@declare_process_block_class("IXParameterBlock")
+class IXParameterData(PhysicalParameterBlock):
+    CONFIG = PhysicalParameterBlock.CONFIG()
+
+    CONFIG.declare(
+        "solute_list",
+        ConfigValue(domain=list, description="List of solute species names"),
+    )
+
+    CONFIG.declare(
+        "diffusivity_data",
+        ConfigValue(
+            default={},
+            domain=dict,
+            description="Dict of solute species names and bulk ion diffusivity data",
+        ),
+    )
+    CONFIG.declare(
+        "mw_data",
+        ConfigValue(
+            default={},
+            domain=dict,
+            description="Dict of component names and molecular weight data",
+        ),
+    )
+
+    CONFIG.declare(
+        "charge", ConfigValue(default={}, domain=dict, description="Ion charge")
+    )
+
+    # reference_cation and reference_anion for use in future prop pack development
+    CONFIG.declare(
+        "reference_cation",
+        ConfigValue(default="H_+", domain=str, description="Reference Cation"),
+    )
+
+    CONFIG.declare(
+        "reference_anion",
+        ConfigValue(default="OH_-", domain=str, description="Reference Anion"),
+    )
+
+    def build(self):
+        """
+        Callable method for Block construction.
+        """
+        super(IXParameterData, self).build()
+
+        self._state_block_class = IXStateBlock
+
+        # phases
+        self.Liq = AqueousPhase()
+
+        # list to hold all species (including water)
+        self.component_list = Set()
+
+        # components
+        self.H2O = Solvent()
+
+        # blank sets
+        self.cation_set = Set()
+        self.anion_set = Set()
+        self.solute_set = Set()
+        self.ion_set = Set()
+        self.reference_cation = self.config.reference_cation
+        self.reference_anion = self.config.reference_anion
```

Comment:
It is not clear to me what these "reference_*" ions are for.

Reply:
Resin manufacturers present selectivity coefficients in terms of a reference ion for the resin (its like saying, the resin has X times affinity for ion A over the reference ion B). In theory, you can get the selectivity coefficient for any combination of target and presaturant ions using this data. These aren't used in this implementation, but I included them in anticipation of future developments and having a large resin database. I will remove them.